File: public/js/stor.js
// Copyright (c) 2019 zenywallet
var Stor = (function() {
  var db;
  function Stor() {
    if(this instanceof Stor) {
      db = localStorage;
    } else {
      return new Stor();
    }
  }

  function buf2hex(buffer) {
    return Array.prototype.map.call(new Uint8Array(buffer), function(x) {return ('00' + x.toString(16)).slice(-2)}).join('');
  }

  function hex2buf(hexstr) {
    if(hexstr.length % 2) {
      throw new Error('no even number');
    }
    return new Uint8Array(hexstr.match(/.{2}/g).map(function(byte) {return parseInt(byte, 16)}));
  }

  Stor.prototype.add_xpub = function(xpub) {
    var xpubs = db['xpubs'];
    if(xpubs) {
      xpubs = JSON.parse(xpubs);
    } else {
      xpubs = [];
    }
    if(!xpubs.includes(xpub)) {
      xpubs.push(xpub);
      db['xpubs'] = JSON.stringify(xpubs);
    }
  }

  Stor.prototype.set_xpubs = function(xpubs) {
    db['xpubs'] = JSON.stringify(xpubs);
  }

  Stor.prototype.get_xpubs = function() {
    var xpubs = db['xpubs'];
    if(xpubs) {
      xpubs = JSON.parse(xpubs);
    } else {
      xpubs = [];
    }
    return xpubs;
  }

  Stor.prototype.del_xpub = function(xpub) {
    var xpubs = db['xpubs'];
    if(xpubs) {
      xpubs = JSON.parse(xpubs);
      var idx = xpubs.indexOf(xpub);
      if(idx >= 0) {
        xpubs.splice(idx, 1);
        db['xpubs'] = JSON.stringify(xpubs);
      }
    } else {
      xpubs = [];
    }
    if(xpubs.length == 0) {
      db.removeItem('xpubs');
    }
  }

  Stor.prototype.del_xpubs = function() {
    db.removeItem('xpubs');
  }

  Stor.prototype.del_all = function() {
    db.clear();
  }

  Stor.prototype.get_salt = function(create_no_exists) {
    var salt = db['salt'];
    if(salt) {
      return hex2buf(salt);
    }

    if(create_no_exists) {
      var buf = new Uint8Array(32);
      crypto.getRandomValues(buf);
      db['salt'] = buf2hex(buf);

      salt = db['salt'];
      if(salt) {
        return hex2buf(salt);
      }
      return null;
    } else {
      return null;
    }
  }

  Stor.prototype.set_shield = function(data) {
    db['shield'] = buf2hex(data);
  }

  Stor.prototype.get_shield = function() {
    return hex2buf(db['shield']);
  }

  Stor.prototype.set_lock_type = function(lock_type) {
    db['locktype'] = lock_type;
  }

  Stor.prototype.get_lock_type = function() {
    return db['locktype'];
  }

  Stor.prototype.set_lang = function(lang) {
    db['lang'] = lang;
  }

  Stor.prototype.get_lang = function() {
    return db['lang'];
  }

  return Stor;
}());

File: src/js/stor.nim
# Copyright (c) 2019 zenywallet

import std/jsffi
import std/macros
import zenyjs
import zenyjs/core
import zenyjs/seed

macro jsPrototype(obj: typed, field: untyped, body: untyped): untyped =
  var protoStr = "#.prototype." & $field & " = #"
  var protoHelper = ident($obj & "_proto_" & $field)
  quote do:
    proc `protoHelper`(obj, body: auto) {.importjs: `protoStr`.}
    `protoHelper`(`obj`, `body`)

var Stor* {.exportc.} = proc() = discard

var localStorage {.importc, nodecl.}: JsObject
var db = if localStorage.isDefined(): localStorage else: JsObject{}
template hex2buf(str: cstring or JsObject): Uint8Array = hexToUint8Array(str.to(cstring))
template buf2hex(uint8Array: Uint8Array or JsObject): cstring = uint8ArrayToHex(uint8Array)

Stor.jsPrototype(add_xpub):
  proc(xpub: JsObject) =
    var xpubs = db["xpubs"]
    if xpubs.to(bool):
      xpubs = JSON.parse(xpubs)
    else:
      xpubs = [].toJs
    if not xpubs.includes(xpub).to(bool):
      xpubs.push(xpub)
      db["xpubs"] = JSON.stringify(xpubs)

Stor.jsPrototype(set_xpubs):
  proc(xpubs: JsObject) =
    db["xpubs"] = JSON.stringify(xpubs)

Stor.jsPrototype(get_xpubs):
  proc(): JsObject =
    var xpubs = db["xpubs"]
    if xpubs.to(bool):
      xpubs = JSON.parse(xpubs)
    else:
      xpubs = [].toJs
    return xpubs

Stor.jsPrototype(del_xpub):
  proc(xpub: JsObject) =
    var xpubs = db["xpubs"]
    if xpubs.to(bool):
      xpubs = JSON.parse(xpubs)
      var idx = xpubs.indexOf(xpub)
      if idx.to(int) >= 0:
        xpubs.splice(idx, 1)
        db["xpubs"] = JSON.stringify(xpubs)
    else:
      xpubs = [].toJs
    if xpubs.length == 0.toJs:
      db.removeItem("xpubs")

Stor.jsPrototype(del_xpubs):
  proc() = db.removeItem("xpubs")

Stor.jsPrototype(del_all):
  proc() = db.clear()

Stor.jsPrototype(get_salt):
  proc(create_no_exists: bool): JsObject =
    var salt = db["salt"]
    if salt.to(bool):
      return hex2buf(salt)

    if create_no_exists:
      var buf = cryptSeedUint8Array(32)
      db["salt"] = buf2hex(buf)

      salt = db["salt"]
      if salt.to(bool):
        return hex2buf(salt)
      return jsNull
    else:
      return jsNull

Stor.jsPrototype(set_shield):
  proc(data: JsObject) =
    db["shield"] = buf2hex(data)

Stor.jsPrototype(get_shield):
  proc(): JsObject =
    return hex2buf(db["shield"])

Stor.jsPrototype(set_lock_type):
  proc(lock_type: JsObject) =
    db["locktype"] = lock_type

Stor.jsPrototype(get_lock_type):
  proc(): JsObject =
    return db["locktype"]

Stor.jsPrototype(set_lang):
  proc(lang: JsObject) =
    db["lang"] = lang

Stor.jsPrototype(get_lang):
  proc(): JsObject =
    return db["lang"]
